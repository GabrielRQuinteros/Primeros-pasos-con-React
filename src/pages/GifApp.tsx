import { useEffect, useState } from "react"
import AddCategory from "../components/AddCategory/AddCategory";
import GifGrid from "../components/GifGrid/GifGrid";
import { getSearchedGifs } from "../utils/GiphyAPI";
import type { Gif } from "../interfaces/GifInterface";

const GifApp = () => {
    const [categories, setCategories] = useState(['One Punch Man', 'Dragon Ball']);
    let gifs: Gif[] = [];
    //! NOTA IMPORTANTE:
    // NUNCA PONER CONDICIONALES QUE AFECTEN A LA CREACIÓN DE HOOKS, VA ENCONTRA DE COMO SE ESPERA QUE SE USEN LOS HOOKS Y TRAE PROBLEMAS.
    // if(true) {
    //     const [categorias, setCategorias] = useState(['One Punch Man', 'Dragon Ball']);
    // }

    useEffect( () => {
        getSearchedGifs('One Punch Man', 5, 0).then( response => {
            gifs = response;
        } ).catch( error => {
            console.log(error);
        } );
    }, []);

    const onAddCategory = ( newCategory: string ) => {
        if( categories.includes( newCategory ) )
            return;

        setCategories( [ newCategory, ...categories ] );
    }

  return (
    <>
        {/* titulo */}
        <h1>Aplicación de Gifs</h1>
        {/* Input */}
            <AddCategory defaultValue=""
                        //  categorias={categorias}
                        //  callback={ setCategorias }
                        onNewValue={ onAddCategory }
                          />
        {/* listado de gifs */}

        <ol>
            {  categories.map( ( cat ) => (
                <GifGrid category={cat}
                         key={cat} 
                    />
                
                ))  
            }
        </ol>

        { gifs.map( gif => ( <img src={gif.url} alt="" /> ) ) }

    </>
  )
}

export default GifApp
