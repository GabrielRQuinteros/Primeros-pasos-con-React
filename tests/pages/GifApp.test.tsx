import { fireEvent, render, screen } from "@testing-library/react";
import GifApp from '../../src/pages/GifApp';

describe('Pruebas del componente <GifApp/>', () => {
  
    const categoryInput = 'Dio';

    test('1-Test-Se prueba que se renderice el componente.', () => {
        const { container }=render( ( <GifApp /> ) );
        expect( container ).toMatchSnapshot();
    });

    test('2-Test-Se prueba que no se puedan duplicar categorias', () => {
        
        // RENDERIZO EL COMPONENTE
        render( ( <GifApp /> ) );

        // OBTENGO EL CAMPO Y EL FORMULARIO PARA REALIZAR LOS SUBMIT
        const input: HTMLInputElement = screen.getByRole('textbox');
        const form : HTMLFormElement = screen.getByRole('form');

        // INGRESO + SUBMIT DE CATEGORIA
        fireEvent.input( input, { target: { value: categoryInput } } );
        fireEvent.submit(form);

        // Vuelvo a buscar la misma categoria
        fireEvent.input( input, { target: { value: categoryInput } } );
        fireEvent.submit(form);
        
        // OBTENER TODOS LOS H3 RENDERIZADOS
        const h3Elements = screen.getAllByRole('heading', { level: 3 });

        // FILTRAR LOS QUE CONTENGAN EL TEXTO DE categoryInput
        const matchingH3s = h3Elements.filter(h3 =>
            h3.textContent?.toLowerCase().includes(categoryInput.toLowerCase())
        );

        // ASSERT: Deber√≠a haber solo uno si no se permiten duplicados
        expect(matchingH3s.length).toBe(1);
    });


    


})
