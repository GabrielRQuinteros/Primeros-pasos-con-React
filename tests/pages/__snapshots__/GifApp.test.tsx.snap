// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <GifApp/> 1-Test-Se prueba que se renderice el componente. 1`] = `
<div>
  <h1>
    Aplicación de Gifs
  </h1>
  <form
    aria-label="form"
  >
    <input
      id="inputCategoria"
      placeholder="Buscar Categoría"
      type="text"
      value=""
    />
  </form>
  <ol>
    <h3>
      One Punch Man
    </h3>
    <h2>
      Cargando ...
    </h2>
    <div
      class="card-grid"
    />
    <h3>
      Dragon Ball
    </h3>
    <h2>
      Cargando ...
    </h2>
    <div
      class="card-grid"
    />
  </ol>
</div>
`;
